File: kotlin/roman-numerals/src/main/kotlin/RomanNumerals.kt
object RomanNumerals {

    private val romanValues = listOf(
        1000 to "M",
        900 to "CM",
        500 to "D",
        400 to "CD",
        100 to "C",
        90 to "XC",
        50 to "L",
        40 to "XL",
        10 to "X",
        9 to "IX",
        5 to "V",
        4 to "IV",
        1 to "I"
    )

    fun value(n: Int): String =
        romanValues.fold(Pair(n, emptyList<String>())) { (remainingNumber, output), (number, romanSymbol) ->
            Pair(remainingNumber % number, output + romanSymbol.repeat(remainingNumber / number))
        }.second.joinToString("")
}

File: typescript/roman-numerals/roman-numerals.ts
const ROMAN_NUMERALS: [number, string][] = [
    [1000, 'M'],
    [900, 'CM'],
    [500, 'D'],
    [400, 'CD'],
    [100, 'C'],
    [90, 'XC'],
    [50, 'L'],
    [40, 'XL'],
    [10, 'X'],
    [9, 'IX'],
    [5, 'V'],
    [4, 'IV'],
    [1, 'I']
];
export const toRoman = (n: number) => {
    let acc = '';
    for (const [val, s] of ROMAN_NUMERALS) {
        acc += s.repeat(n / val)
        n %= val;
    }

    return acc;
};
